drivers/Aqara/aqara-lock/src/init.lua

```diff
@@ -0,0 +1,233 @@
+local CLOUD_PUBLIC_KEY = "__cloud_public_key"
+
+local function my_secret_data_handler(driver, device, secret_info)
+  -- At time of writing this returns nothind beyond "secret_type = aqara"
+  local shared_key = secret_info.shared_key
+  local cloud_public_key = secret_info.cloud_public_key
+
+  device:set_field(SHARED_KEY, shared_key, { persist = true })
+  device:set_field(CLOUD_PUBLIC_KEY, cloud_public_key, { persist = true })
+  credential_utils.save_data(driver)
+
+  if cloud_public_key ~= nil then
+    local raw_data = base64.decode(cloud_public_key)
+    -- send cloud_pub_key
+    device:send(cluster_base.write_manufacturer_specific_attribute(device,
+      PRI_CLU, PRI_ATTR, MFG_CODE, data_types.OctetString, "\x3E" .. raw_data))
+  end
+end
+
+local function device_added(self, device)
+  if credential_utils.is_exist_host(device) then
+    device:emit_event(remoteControlStatus.remoteControlEnabled('true', { visibility = { displayed = false } }))
+  else
+    credential_utils.set_host_count(device, 0)
+    device:emit_event(remoteControlStatus.remoteControlEnabled('false', { visibility = { displayed = false } }))
+  end
+  device:emit_event(Battery.battery(100))
+  device:emit_event(TamperAlert.tamper("clear", { visibility = { displayed = false } }))
+  device:emit_event(Lock.lock("locked"))
+  credential_utils.save_data(self)
+end
+
+local function toValue(payload, start, length)
+  local ret = 0
+  for i = start, start + length - 1 do
+    ret = (ret << 8) + string.byte(payload, i)
+  end
+  return ret
+end
+
+local function toHex(value, length)
+  local ret = string.char(0xFF & value)
+  for i = length, 2, -1 do
+    ret = string.char(0xFF & (value >> 8 * (i - 1))) .. ret
+  end
+  return ret
+end
+
+local function event_lock_handler(driver, device, evt_name, evt_value)
+  if evt_value == 1 then
+    device:emit_event(Lock.lock(evt_name))
+    device:emit_event(TamperAlert.tamper.clear())
+    if credential_utils.is_exist_host(device) then
+      device:emit_event(remoteControlStatus.remoteControlEnabled('true', { visibility = { displayed = false } }))
+    else
+      device:emit_event(remoteControlStatus.remoteControlEnabled('false', { visibility = { displayed = false } }))
+    end
+  end
+end
+
+local function event_unlock_handler(driver, device, evt_name, evt_value)
+  local id, label
+  id, label = credential_utils.find_userLabel(driver, device, evt_value)
+  device:emit_event(Lock.lock.unlocked({ data = { method = evt_name, codeId = id, codeName = label } }))
+  device:emit_event(remoteControlStatus.remoteControlEnabled('false', { visibility = { displayed = false } }))
+  device:emit_event(TamperAlert.tamper.clear())
+end
+
+local function event_door_handler(driver, device, evt_name, evt_value)
+  if evt_value == 2 then
+    device:emit_event(Lock.lock(evt_name))
+    device:emit_event(TamperAlert.tamper.clear())
+  elseif evt_value == 4 then
+    device:emit_event(TamperAlert.tamper.detected())
```

Comment:
just checking: should a `Lock.lock("not fully locked")` event not be emitted here?

Reply:
I have added it.

Reply:
`evt_value=4` means "Door Forced Open"

Reply:
Sorry, I misunderstood.